dart/lib/src/recursive_exception_cause_extractor.dart

```diff
@@ -17,7 +17,7 @@ class RecursiveExceptionCauseExtractor {
 
     var currentException = exception;
     ExceptionCause? currentExceptionCause =
-        ExceptionCause(exception, stackTrace);
+        ExceptionCause(exception, stackTrace ?? (exception is Error) ? (exception as Error).stackTrace : null);
```

Comment:
I think this would also be a nice addition in general for all Dart/Flutter `Error` sub-classes.

Reply:
Yeah but we do that already.
https://github.com/getsentry/sentry-dart/blob/f1314d57fef996236ceca00725cc9c3acce16d85/dart/lib/src/sentry_exception_factory.dart#L27-L29
I guess we just don't do that when its a chained exception.

Reply:
There is, your screenshot is a chained exception, the String has a stack trace, but the DioError does not and your fix should solve this, I guess.
`DioError` is an `Exception` and not an `Error` that's why it didn't work automatically.

Reply:
> `DioError` is an `Exception` and not an `Error` that's why it didn't work automatically.

****, yea you are right. Would it be generally useful to cast the error `Object` to `dynamic` and check if it has useable `stackTrace` property? Right there in the `SentryExceptionFactory` I mean, independent of the Dio case here?

Reply:
I don't think so, if the exception is an `Error` we already do that (it's a known type).
Any other type/chained exception type has to do that beforehand using the extractor, in this case specifically, we just need to fix the extractor IMO.
Unless having a `stackTrace` property in a custom exception is a well-known and used pattern for the most custom exceptions, I don't think it's the case.

Reply:
But then everything is correct, this is what is currently being done. But it is unusable.
Every `DioError` without a cause (`DioError.error`) is named `DioError inner stacktrace` because this is then extract as the cause.

A stackTrace is not a cause, it just belongs to the original error.

Reply:
What if we add a ` StackTrace? stackTrace(T error) => null;` function to `ExceptionCauseExtractor` and use this here for every error: https://github.com/getsentry/sentry-dart/blob/f1314d57fef996236ceca00725cc9c3acce16d85/dart/lib/src/sentry_exception_factory.dart#L27-L29 and fallback to the current behavior if it is still null afterwards.
This can then be overridden for each error type where it is required.

Reply:
> But then everything is correct, this is what is currently being done. But it is unusable.
> Every DioError without a cause (DioError.error) is named DioError inner stacktrace because this is then extract as the cause.
> 
> A stackTrace is not a cause, it just belongs to the original error.

Nope, that's not how it is, your PR fixes it and makes it like this.

Reply:
I have to run the code and check how it is and the changes, right now I'm trying to understand based on only looking into the code, I'm not aware of the dio v5 changes as well besides reviewing the migration PR.
I will check this carefully next week and come back to you, sorry about the back and forth.

Reply:
> > But then everything is correct, this is what is currently being done. But it is unusable.
> > Every DioError without a cause (DioError.error) is named DioError inner stacktrace because this is then extract as the cause.
> > A stackTrace is not a cause, it just belongs to the original error.
> 
> Nope, that's not how it is, your PR fixes it and makes it like this.

No worries, with my PR there is no stacktrace at all when there is no `DioError.error`. There is no code that extracts `DioError.stackTrace` in that case.

Reply:
As mentioned before, this change specifically isn't needed cus https://github.com/getsentry/sentry-dart/blob/f1314d57fef996236ceca00725cc9c3acce16d85/dart/lib/src/sentry_exception_factory.dart#L27-L29 does it automatically.
The problem is that `DioError` isn't an `Error` so the `DioError.stackTrace` is never read in case the given `stackTrace` param is null, ideally we'd fallback to `DioError.stackTrace`.
I believe this should be solved by `ExceptionCause`.
Like a way to resolve a better `stackTrace` from the error for known types, not generally from the cause, can be from the root cause as well.

Reply:
Your suggestion of parsing it to `dynamic` and accessing the `stackTrace` would work in this case but its not scalable because you might have an Exception that has a `stackTrace` field called differently.
ideally we'd have something similar to the `ExceptionCause` but for stack trace.

Such as:

```dart
abstract class ExceptionCauseExtractor<T> {
  ExceptionCause? cause(T error);
  dynamic stackTrace(T error);
  Type get exceptionType => T;
}
```

`DioErrorExtractor` then:

```dart
@override
  dynamic stackTrace(DioError error) => error.stackTrace;
```

And later the `RecursiveExceptionCauseExtractor` does:

```dart
final betterStackTrace = stackTrace ?? extractor?.stackTrace(extractionSourceSource);
```

And use `betterStackTrace` as the `stackTrace` instead of something else.

How does that sound?

Reply:
Maybe this should not be part of `ExceptionCauseExtractor` cus the `stackTrace` isn't related to the cause at all, but it was easy to hack it up.

Reply:
`ExceptionStackTraceExtractor` sounds good.

> Not sure, I think the manually provided stackTrace should have precedence.

Most likely there will be cases where the provided stackTrace is better or worse than the extracted one.
So that's fine, I'd expect that people don't provide an `ExceptionStackTraceExtractor` for a known type if the provided stack trace is already enough, so it's a fair trade-off.

Would you like to adapt your PR to consider `ExceptionStackTraceExtractor`? a separate PR is also fine.

Reply:
Yea I'll start working on a new PR.